Zypper repository rewrite

robinro opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
packaging/os/zypper_repository.py

##### ANSIBLE VERSION
devel

##### SUMMARY
This continues the rewrite proposed in #1515. But expand it:
* add properties `enabled` and `priority`
* allow changing of one property and correctly report changed
* allow overwrite of multiple repositories by alias and URL
* cleanup of unused code and more structuring

* fixes #1515 (rewrite this is based on)
* fixes #939 (change repo url) 
* fixes #1633 (change auto-refresh)
* fixes #1450 (add prio)

gregdek commented:
Thanks @robinro. To the current maintainers, @matze please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @robinro for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

robinro commented:
ready_for_review

gregdek commented:
Thanks @robinro. To the current maintainers, @matze please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

matze commented:
In principle I'd say shipit, however the removal of support for older zypper versions in 58f1021 should be communicated appropriately.

gregdek commented:
Thanks again to @robinro for this PR, and thanks @matze for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

robinro commented:
Thanks @matze 

Do you know how old zypper must be to require the old setup?
According to
https://en.opensuse.org/SDB:Zypper_usage_11.3#XML_output
the xml output was there in 11.3 already. Then also SLES 11.2 should have probably had it.

robinro commented:
According to https://en.opensuse.org/openSUSE:Zypper_versions you have to go to opensuse 11.0 to have the old zypper version (0.x). In SLES I unterstand that 11.\* is fine, see https://www.suse.com/LinuxPackages/packageRouter.jsp?product=server&version=11&architecture=i386&package_name=index_all

@paavoap: Your my go-to guy for old SLES versions :) This PR would remove support for zypper 0.x. Does one of your older SLES versions still use it? Can you confirm that SLES 11 is already on zypper 1.x?

I added a preliminary note to the requirements. It should be updated once we know the SLES version needed.

matze commented:
> Do you know how old zypper must be to require the old setup?

That does not matter if it breaks at least one person's setup. As I said, I am not against that change itself but it should be communicated appropriately.

robinro commented:
@matze is a note in the requirements enough to communicate it?

matze commented:
I guess that's enough.

Robin Roth notifications@github.com schrieb am Do., 14. Apr. 2016, 08:06:

> @matze https://github.com/matze is a note in the requirements enough to
> communicate it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-extras/pull/1990#issuecomment-209776181

paavoap commented:
@robinro Fresh install from DVD

```
# cat /etc/SuSE-release
SUSE Linux Enterprise Server 11 (x86_64)
VERSION = 11
PATCHLEVEL = 0
# zypper --version
zypper 1.0.8
```

robinro commented:
@paavoap Thanks! So this aligns with my guess in bad0162

alxgu commented on packaging/os/zypper_repository.py:
```diff
@@ -58,16 +58,30 @@
+        version_added: "2.1"
+requirements: 
+    - "zypper  >= 1.0"
+    - "openSuSE >= 11.1"
```
opensuse or sles are not really requirements. That should be hints.
- "zypper >= 1.0" # included in openSuSE >= 11.1 or SuSE Linux Enterprise Server/Desktop >= 11.0

robinro replied:
sounds reasonable, added it

resmo commented:
I am not able the tests but based on @robinro comments on IRC this PR is tested and tests extended in https://github.com/ansible/ansible/pull/15387.

robinro commented:
integration test output

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
Monday 18 April 2016  17:51:07 +0200 (0:00:00.095)       0:00:00.095 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Delete] *****************************************
Monday 18 April 2016  17:51:07 +0200 (0:00:00.533)       0:00:00.629 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : Add repo] ***************************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.289)       0:00:00.918 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : debug] ******************************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.300)       0:00:01.219 *********\* 
ok: [myhostname] => {
    "zypper_result": {
        "changed": true, 
        "repodata": {
            "alias": "test", 
            "autorefresh": "1", 
            "enabled": "1", 
            "gpgcheck": "1", 
            "name": null, 
            "priority": null, 
            "url": "http://dl.google.com/linux/chrome/rpm/stable/x86_64"
        }, 
        "state": "present"
    }
}

TASK [test_zypper_repository : verify repo addition] ***************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.082)       0:00:01.302 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Add repo again] *********************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.074)       0:00:01.377 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : verify no change on second install] *************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.292)       0:00:01.670 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Change repo URL] ********************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.072)       0:00:01.743 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : Verify change on URL only change] ***************
Monday 18 April 2016  17:51:09 +0200 (0:00:00.423)       0:00:02.166 *********\* 
ok: [myhostname]

PLAY RECAP *********************************************************************
myhostname        : ok=9    changed=3    unreachable=0    failed=0

resmo commented:
great @robinro thanks for the test results
